Provide access to method that handles commands

xPanini opened the pull request:
**Commit Details**
-Move startDedInputCallBack functionality to ExecuteCommand method
-Update startDedInputCallBack method to use ExecuteCommand
-Add ExecuteCommand method which takes a string and CommandCaller for allowing execution of commands from other areas

### Description of the Change
startDedInputCallBack is a method used to accept commands on the command console asynchronously. It takes no parameters for outside use.

This change moves the code inside startDedInputCallBack to its own method named ExecuteCommand. ExecuteCommand accepts two parameters, text input and a CommandCaller object, to allow executing commands from other areas.

### Alternate designs
Another design was to add two optional parameters, input = null and CommandCaller = null. I moved away from this because it was clunky, requiring escaping the while loop in the method, and having an unclear method name.

### Why this should be merged into tModLoader
This will allow mod developers to gain access to the command executor. Different possibilities include web APIs, RESTful APIs, or for alternative means of executing commands (such as a Discord relay to run commands in chat).

### Benefits
Access to the command executor for all mod developers

### Possible drawbacks
N/A. The method is optional and the former method has kept it's signature and functionality.

### Applicable Issues
The text variable is slightly unclear, not indicating how to use the method. I did not rename the text variable as that was formerly included as a variable named text that accepted Console input. This limited the amount of changes to use, but also reduced the patch size to rename the text variable. 

### Sample Usage
An example CommandSign would be an excellent idea. A sign that takes the text on the sign and runs the command when right clicked on or something. 
ExecuteMethod("noon", new ConsoleCommand()); 
ExecuteMethod("item Panini bomb 30", new CustomCommandCaller());

Agrair commented:
I don't understand, is this a way to manage when commands are executed or manually execute commands?

xPanini commented:
It is a way to manually execute commands from within code. The "text" variable accepts input like "noon" or "giveitem Panini Bomb" and it will execute the command.

This is only possible with reflection currently and makes it difficult to make a mod that executes commands.

One example is my TerrariaChatRelay mod to run commands from Discord. Another is a Command blocks mod like Minecraft.

Chicken-Bones commented:
You've done an excellent job keeping the diffs small. Add that ExampleCommandSign and I'll merge it

xPanini commented:
Added ExampleCommandSign. Just an FYI, I have added a patch to the WorldIO and WorldFile classes to add support for ModSigns. Without it, the sign will not save to the world.

Chicken-Bones commented on ExampleMod/Tiles/ExampleCommandSign.cs:
```diff
@@ -0,0 +1,115 @@
+				string[] array = text.Split('\n');
+				foreach (string value in array)
+				{
+					Main.NewText(value);
```
Could be a 2-liner with RemoveEmptyEntries in the Split

Chicken-Bones commented on patches/tModLoader/Terraria.IO/WorldFile.cs.patch:
```diff
@@ -17,12 +18,13 @@
  						memoryStream.Close();
 +						WorldHooks.SetupWorld();
 +						WorldIO.Load(Main.worldPathName, flag);
++
```
Accidental newline

Chicken-Bones commented on patches/tModLoader/Terraria.IO/WorldFile.cs.patch:
```diff
@@ -1,11 +1,12 @@
  using Terraria.Utilities;
 +using Terraria.ModLoader;
 +using Terraria.ModLoader.IO;
++using System.Windows.Forms;
```
Accidental import?

Chicken-Bones commented on patches/tModLoader/Terraria.IO/WorldFile.cs.patch:
```diff
@@ -150,6 +161,23 @@
+-					sign = null;
+ 				}
+ 				Main.sign[i] = sign;
+ 			}
```
I'll need to look over save load in more depth later

xPanini replied:
Adjusted commit via conversation in PM

Jofairden commented on patches/tModLoader/Terraria.IO/WorldFile.cs.patch:
```diff
@@ -150,6 +150,23 @@
+ 				Tile tile = Main.tile[num2, num3];
+ 				Sign sign;
+-				if (tile.active() && Main.tileSign[(int)tile.type])
++				if (true) //if (tile.active() && Main.tileSign[(int)tile.type]) 
```
I'm not sure about this, it looks very volatile.

Chicken-Bones replied:
I've approved it

Jofairden commented on ExampleMod/Tiles/ExampleCommandSign.cs:
```diff
@@ -0,0 +1,108 @@
+		public override void RightClick(int i, int j)
+		{
+			// Uses the text from the sign to run a command
+			Main.ExecuteCommand(Main.sign[Sign.ReadSign(i, j, true)].text, new ExampleCommandCaller());
```
Just making sure, this will work fine if the text passed isn't a valid command?

xPanini replied:
It works fine; there is a try/catch block around the all the logic that will catch invalid commands.

Jofairden commented on ExampleMod/Tiles/ExampleCommandSign.cs:
```diff
@@ -0,0 +1,108 @@
+{
+	public class ExampleCommandSign : ModTile
+	{
+		public ExampleCommandSign()
```
Redundant constructor, remove it.

Jofairden commented on ExampleMod/Items/Placeable/ExampleCommandSign.cs:
```diff
@@ -0,0 +1,37 @@
+	{
+		public override void SetStaticDefaults()
+		{
+			Tooltip.SetDefault("This is a modded sign.");
```
Maybe make it more descriptive, like: "Enter a valid command on this sign to execute it!"

Jofairden commented on ExampleMod/Tiles/ExampleCommandSign.cs:
```diff
@@ -0,0 +1,108 @@
+
+			public Player Player => null;
+
+			public void Reply(string text, Color color = default(Color))
```
Is Reply not an `override`? How will modders know to use this method?

Jofairden replied:
My bad, it's an interface 👍

Jofairden commented on ExampleMod/Tiles/ExampleCommandSign.cs:
```diff
@@ -0,0 +1,108 @@
+		{
+			public CommandType CommandType => CommandType.Console;
+
+			public Player Player => null;
```
Idem dito for the properties about virtual/abstract override

Jofairden replied:
My bad, it's an interface 👍

Chicken-Bones commented on patches/tModLoader/Terraria.IO/WorldFile.cs.patch:
```diff
@@ -150,6 +150,23 @@
+ 					sign.y = num3;
+ 				}
+-				else
++				else 
```
Nice space :')

xPanini replied:
Reeeee I was so careful to check lmao!!!!!!

Chicken-Bones commented on patches/tModLoader/Terraria/NetMessage.cs.patch:
```diff
@@ -245,6 +245,15 @@
+ 								}
+ 							}
+-							if (tile2.type == 425 && tile2.frameX % 36 == 0 && tile2.frameY % 36 == 0)
++							if ((tile2.type == 425 && tile2.frameX % 36 == 0 && tile2.frameY % 36 == 0) || Main.tileSign[tile2.type])
```
This looks like the vanilla logic only sends the sign text for one corner of the tile. Should modded signs be different?

xPanini replied:
Hovering over all tiles of the sign, the text of the sign is shown even in multiplayer. This leads me to believe that all tiles share the data of the corner of the tile.

xPanini replied:
I take back what I said. It sends the sign data for each sign tile there is when it should only send data for the one corner. I need to think about this for a while.

GoodPro712 commented:
Can't devs commit to the pr themself so this can be merged? Panini has left modding so is unable to fix this

Chicken-Bones commented:
If another modder wants this feature, we're happy for someone else to take it on, they make a branch which tracks xPanini/tModLoader/ExecuteCommand.

Since no-one has bumped the issue we assume it's not a priority feature

GoodPro712 commented:
How would I make a branch which tracks xPanini/tModLoader/ExecuteCommand.

xPanini commented:
I added the last line of code this needed to work for 36 x 36 signs (the size of regular signs). 
While this doesn't allow for any other size than 36 x 36, it's better than nothing and can be extended in the future. 

The last thing that needs to be done is to **merge this with the current tModLoader version** as this is **very far behind** (This is based on version 0.11).

If any of the tMod devs need a quick reminder or run-down of what's going on, feel free to ping me on Discord.

Cheers. 🍻

Plantterror commented:
Interested in this getting developed a lot more. I have a near mod idea that could take advantage of this.

xPanini commented:
Thanks a bunch!
